### PR TITLE
T/296 Build Converters

### LINKED ISSUE #296: Converter Builders
## View To Model:

It is not that simple to build an flexible helper functions for converters with readable API, they take too many parameters. But it should be simple to build converters.

This is why I suggest chainable API:

```js
BuildViewConverterUsing( editor.data.toModel, editor.editing.toModel )
	.fromElement( 'strong' )
	.fromElement( 'b' )
	.fromElement( { element: 'span', style: { 'font-weight': fontWeightBold } } ).consuming( { style: 'font-weight' } )
	.toAttribute( { bold: true } );

BuildViewConverterUsing( editor.data.toModel, editor.editing.toModel )
	.fromElement( 'img' ).consuming( { name: true, class: 'src' } )
	.toElement( ( viewElement ) => new ModelElement( 'img', { 'src': viewElement.getAttribute( 'src' ) } ) );

BuildViewConverterUsing( editor.data.toModel, editor.editing.toModel )
	.fromElement( 'p' )
	.toElement( ( viewElement ) => new ModelElement( 'p' ) );

BuildViewConverterUsing( editor.data.toModel, editor.editing.toModel )
	.fromAttribute( { class: 'important' } )
	.toAttribute( { important: true } );

BuildViewConverterUsing( editor.data.toModel, editor.editing.toModel )
	.fromElement( 'a' ).consuming( { name: true, attribute: 'href' } )
	.toAttribute( ( viewElement ) => { return { 'link': viewElement.getAttribute( 'href' ) }; } );
```

Note that:
 - `BuildViewConverterUsing()` may take many `ViewConversionDispatcher`s, so it may handle many pipelines (editing and data) at the same time, or user can build separate converters for each pipeline.
 - `fromElement` take `Matcher` (https://github.com/ckeditor/ckeditor5-engine/issues/274) instance or  string or object which will be used to build `Matcher`.
 - it is possible to add multiple elements listeners using `fromElement`,
 - `consuming` defines the list of elements to consume from the last added element, or, if not defined, the consuming list is created based on the match,
 - the API is scalable, is should be possible to add more methods in the future if needed (i.e. `.withoutChildren()`).

## Model To View:

Converters from model to view could provide similar API:

```js
BuildModelConverterUsing( editor.data.toView, editor.editing.toView )
	.fromElement( 'p' )
	.toElement( ( modelElement ) =>  new ViewElement( 'p' ) );

BuildModelConverterUsing( editor.data.toView, editor.editing.toView )
	.fromAttribute( 'link' )
	.toElement( ( key,  value ) =>  new ViewElement( 'a', { href: value } ) );

BuildModelConverterUsing( editor.data.toView, editor.editing.toView )
	.fromAttribute( 'bold' )
	.toElement( 'strong' );
```

### PR DESCRIPTION
**Pull request by scofalik**
Fixes #296 

There are still no attributes priorities as this is waiting for #372.

### CODE REVIEW

File: src/treecontroller/model-converter-builder.js [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+		return this;
+	}
+
+	/**
+	 * Changes default priority for built converter.
+	 *
+	 * @param {Number} priority Converter priority.
+	 * @returns {engine.treeController.ModelConverterBuilder}
+	 */
+	withPriority( priority ) {
```
Comment (pjasiun) [inline src/treecontroller/model-converter-builder.js]: Sample would be useful.

File: src/treecontroller/view-converter-builder.js [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+	 *			.toAttribute( { bold: true } );
+	 *
+	 *		BuildViewConverterFor( dispatcher )
+	 *			.fromElement( 'img' ).consuming( { name: true, attributes: [ 'src', 'title' ] } )
+	 *			.toElement( ( viewElement ) => new ModelElement( 'image', { src: viewElement.getAttribute( 'src' ),
+	 *																		title: viewElement.getAttribute( 'title' ) } );
+	 *
+	 * **Note:** All and only values from passed object has to be consumable on converted view element. This means that
+	 * using `consuming` method, you can either make looser conversion conditions (like in first example) or tighter
+	 * conversion conditions (like in second example). Nevertheless, the view element has to match `from...` query clause.
```
Comment (pjasiun) [inline src/treecontroller/view-converter-builder.js]: I do not understand last sentence here.

File: src/treecontroller/model-converter-builder.js (hunk 1 above)
Comment (pjasiun) [inline src/treecontroller/model-converter-builder.js]: Also this is both event priority and attribute priority, what should be mentioned.

File: src/treecontroller/model-to-view-converters.js [hunk 3]
```diff
@@ -218,6 +218,14 @@ export function wrap( elementCreator ) {
 			elementCreator.clone( true ) :
 			elementCreator( data.attributeNewValue, data, consumable, conversionApi );
 
+		// If this is a change event (because old value is not empty) and the creator is a function (so
+		// it may create different view elements basing on attribute value) we have to create
+		// view element basing on old value and unwrap it before wrapping with a newly created view element.
+		if ( data.attributeOldValue !== null && !( elementCreator instanceof ViewElement ) ) {
+			const oldViewElement = elementCreator( data.attributeOldValue, data, consumable, conversionApi );
+			conversionApi.writer.unwrap( viewRange, oldViewElement, evt.priority );
```
Comment (pjasiun) [inline src/treecontroller/model-to-view-converters.js]: This means we do not need to introduce `rewrap` anymore?

Comment (scofalik) [inline src/treecontroller/model-to-view-converters.js]: Yes.

File: src/treemodel/schema.js [hunk 4]
```diff
@@ -365,6 +365,8 @@ export default class Schema {
 		// If attributes property is a string or undefined, wrap it in an array for easier processing.
 		if ( !isArray( query.attributes ) ) {
 			query.attributes = [ query.attributes ];
+		} else if ( query.attributes.length === 0 ) {
+			query.attributes.push( undefined );
```
Comment (pjasiun) [inline src/treemodel/schema.js]: Say WAT?

**Comment by pjasiun** [comment]
After documentation updated this ticket can be merged. Good job!

File: src/treecontroller/model-converter-builder.js (hunk 1 above)
Comment (scofalik) [inline src/treecontroller/model-converter-builder.js]: Added

File: src/treecontroller/view-converter-builder.js (hunk 2 above)
Comment (scofalik) [inline src/treecontroller/view-converter-builder.js]: I rewritten it in (hopefully) more understandable way.

**Pull request merged**